starlette/routing.py

```diff
@@ -56,22 +56,38 @@ def iscoroutinefunction_or_partial(obj: typing.Any) -> bool:  # pragma: no cover
     return inspect.iscoroutinefunction(obj)
 
 
+def _to_async(
+    func: typing.Callable[[Request], typing.Awaitable[Response] | Response],
+) -> typing.Callable[[Request], typing.Awaitable[Response]]:
+    """
+    Takes a function or coroutine `func(request) -> response`,
+    and returns an async function to process the request.
+    """
+    if is_async_callable(func):
+        return func
+
+    def wrapped_func(req: Request) -> typing.Awaitable[Response]:
+        return run_in_threadpool(
+            typing.cast(typing.Callable[[Request], Response], func), req
+        )
+
+    return wrapped_func
+
+
 def request_response(
     func: typing.Callable[[Request], typing.Awaitable[Response] | Response],
 ) -> ASGIApp:
     """
     Takes a function or coroutine `func(request) -> response`,
     and returns an ASGI application.
     """
+    f = _to_async(func)
```

Comment:
Can we drop the diff to this?
```suggestion
    f = func if is_async_callable(func) else functools.partial(run_in_threadpool, func)
```

Reply:
This does not work because in mypy, `functools.partial` is typed as an object. See https://github.com/python/mypy/issues/1484.

The reason I added the helper function was to work around the problems in the typing signature, as `is_async_callable` is incorrectly returning a function of type `AwaitableCallable[Any]` and furthermore typeguards do not work for the negative case in Python.

Reply:
Can we just add a `# type: ignore`?

Reply:
Done